pkg/controllers/gateway/test_helper.go

```diff
@@ -0,0 +1,212 @@
+	if subDomain == failFetchDNSsubdinain {
+		return &v1alpha1.DNSRecord{}, fmt.Errorf(failFetchDNSsubdinain)
+	}
+	record := &v1alpha1.DNSRecord{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      managedZone.Spec.DomainName,
+			Namespace: managedZone.GetNamespace(),
+		},
+	}
+
+	if err := h.controlClient.Get(ctx, client.ObjectKeyFromObject(record), record); err != nil {
+		return nil, err
+	}
+	return record, nil
+}
+
+func (h *TestHostService) GetManagedZoneForHost(ctx context.Context, host string, t traffic.Interface) (*v1alpha1.ManagedZone, string, error) {
+	hostParts := strings.SplitN(host, ".", 2)
+	if len(hostParts) < 2 {
+		return nil, "", fmt.Errorf("unable to parse host : %s on traffic resource : %s", host, t.GetName())
+	}
+	subDomain := hostParts[0]
+	zone := v1alpha1.ManagedZoneList{}
+
+	err := h.controlClient.List(ctx, &zone, client.InNamespace(t.GetNamespace()))
+	if err != nil {
+		return nil, "", err
+	}
+	if len(zone.Items) == 0 {
+		return nil, subDomain, nil
+	}
+	return &zone.Items[0], subDomain, nil
+}
+
+func (h *TestHostService) AddEndpoints(_ context.Context, gateway traffic.Interface, _ *v1alpha1.DNSRecord) error {
+	hosts := gateway.GetHosts()
+	for _, host := range hosts {
+		if host == failEndpointsHostname {
+			return fmt.Errorf(failEndpointsHostname)
+		}
+	}
+	return nil
+}
+
+func newTestHostService(client client.Client) *TestHostService {
+	return &TestHostService{controlClient: client}
+}
+
+type TestGatewayPlacer struct{}
+
+func (p *TestGatewayPlacer) Place(_ context.Context, upstream *v1beta1.Gateway, _ *v1beta1.Gateway, _ ...v1.Object) (sets.Set[string], error) {
+	if upstream.Labels == nil {
+		return nil, nil
+	}
+	if *upstream.Spec.Listeners[0].Hostname == failPlacementHostname {
+		return nil, fmt.Errorf(failPlacementHostname)
+	}
+	targetClusters := sets.Set[string](sets.NewString())
+	targetClusters.Insert(testCluster)
+	return targetClusters, nil
+}
+
+func (p *TestGatewayPlacer) GetPlacedClusters(_ context.Context, gateway *v1beta1.Gateway) (sets.Set[string], error) {
+	if gateway.Labels == nil {
+		return nil, nil
+	}
+	placedClusters := sets.Set[string](sets.NewString())
+	placedClusters.Insert(testCluster)
+	return placedClusters, nil
+}
+
+func (p *TestGatewayPlacer) GetClusters(_ context.Context, _ *v1beta1.Gateway) (sets.Set[string], error) {
+	return nil, nil
+}
+
+func (p *TestGatewayPlacer) ListenerTotalAttachedRoutes(_ context.Context, _ *v1beta1.Gateway, listenerName string, _ string) (int, error) {
+	if listenerName == validHostname {
+		return 1, nil
+	}
+	return 0, nil
+}
+
+func (p *TestGatewayPlacer) GetAddresses(_ context.Context, _ *v1beta1.Gateway, _ string) ([]v1beta1.GatewayAddress, error) {
+	t := v1beta1.IPAddressType
+	return []v1beta1.GatewayAddress{
+		{
+			Type:  &t,
+			Value: "1.1.1.1",
+		},
+	}, nil
+}
+
+func newTestGatewayPlacer() *TestGatewayPlacer {
+	return &TestGatewayPlacer{}
+}
```

Comment:
This file is being included as part of the pkg and not a test file, so it's skewing your coverage report. You should probably move it to a `test` package, or fgure out how you can include test helpers for unit tests.

Reply:
well, in fact, it bumps coverage up right now :laughing:  but I see the point. 
Why don't we take it even a bit further - I moved all that is not the test cases into the `test` package and once integration tests are done they'd end up in the same `test` package?